api/app/src/providers/dexcom.ts

```diff
@@ -0,0 +1,178 @@
+import Provider, { ConsumerHealthDataType } from "./provider";
+import { Config } from "../shared/config";
+import { ConnectedUser } from "../models/connected-user";
+import { updateProviderData } from "../command/connected-user/save-connected-user";
+import { getProviderDataFromConnectUserOrFail } from "../command/connected-user/get-connected-user";
+import { mapToBiometrics } from "../mappings/dexcom/biometrics";
+import { mapToNutrition } from "../mappings/dexcom/nutrition";
+import { dexcomEvgsResp } from "../mappings/dexcom/models/evgs";
+import { dexcomEventsResp } from "../mappings/dexcom/models/events";
+
+const axios = Axios.create();
+
+export class Dexcom extends Provider implements OAuth2 {
+  static URL =
+    Config.isProdEnv() || Config.isSandbox()
+      ? "https://api.dexcom.com/"
+      : "https://sandbox-api.dexcom.com";
+  static AUTHORIZATION_PATH = "/v2/oauth2/login";
+  static TOKEN_PATH = "/v2/oauth2/token";
+  static scopes = "offline_access";
+
+  private static clientId = Config.getDexcomClientId();
+  private static clientSecret = Config.getDexcomClientSecret();
+
+  constructor(
+    private readonly oauth = new OAuth2DefaultImpl(
+      PROVIDER_DEXCOM,
+      Dexcom.clientId,
+      Dexcom.clientSecret,
+      {
+        authorizeHost: Dexcom.URL,
+        tokenHost: Dexcom.URL,
+        authorizePath: Dexcom.AUTHORIZATION_PATH,
+        tokenPath: Dexcom.TOKEN_PATH,
+      },
+      Dexcom.scopes
+    )
+  ) {
+    super({
+      [ConsumerHealthDataType.Activity]: false,
+      [ConsumerHealthDataType.Body]: false,
+      [ConsumerHealthDataType.Biometrics]: true,
+      [ConsumerHealthDataType.Nutrition]: true,
+      [ConsumerHealthDataType.Sleep]: false,
+      [ConsumerHealthDataType.User]: false,
+    });
+  }
+
+  private async getAccessToken(connectedUser: ConnectedUser): Promise<string> {
+    const providerData = getProviderDataFromConnectUserOrFail(connectedUser, PROVIDER_DEXCOM);
+
+    const token = providerData.token;
+
+    const refreshedToken = await this.checkRefreshToken(token, connectedUser);
+
+    return refreshedToken.access_token;
+  }
+
+  async checkRefreshToken(token: string, connectedUser: ConnectedUser): Promise<Token> {
+    const access_token = JSON.parse(token);
+    const isExpired = access_token.expires_at * 1000 - (Date.now() + 120 * 1000) <= 0;
+
+    if (isExpired) {
+      const formData = {
+        grant_type: "refresh_token",
+        refresh_token: access_token.refresh_token,
+        redirect_uri: this.oauth.getRedirectUri(),
+        client_id: Dexcom.clientId,
+        client_secret: Dexcom.clientSecret,
+      };
+
+      const formBody = new URLSearchParams(formData).toString();
+
+      try {
+        const response = await axios.post(`${Dexcom.URL}${Dexcom.TOKEN_PATH}`, formBody, {
+          headers: {
+            "Content-Type": "application/x-www-form-urlencoded",
+          },
+        });
+
+        response.data.expires_at = dayjs(Date.now())
```

Comment:
IIRC we don't need to send `Date.now()` to `dayjs()`, it already initializes with the current timestamp.